iree/tools/iree-e2e-matmul-test.c

```diff
@@ -39,10 +49,198 @@ static iree_status_t iree_get_buffer_view_list_item(
+                                IREE_HAL_MAPPING_MODE_PERSISTENT, access, 0,
+                                IREE_WHOLE_BUFFER, &mapping));
+  *data = mapping.contents.data;
+  return iree_ok_status();
+}
+
+// Allocates host-local |dst| to have the same shape as |src|.
+// Implicitly zero-filled.
+static iree_status_t allocate_host_buffer_view_like(
+    iree_hal_allocator_t* hal_allocator, iree_hal_buffer_view_t* src,
+    iree_hal_buffer_view_t** dst) {
+  return iree_hal_buffer_view_allocate_buffer(
+      hal_allocator, iree_hal_buffer_view_shape_dims(src),
+      iree_hal_buffer_view_shape_rank(src),
+      iree_hal_buffer_view_element_type(src),
+      iree_hal_buffer_view_encoding_type(src),
+      (iree_hal_buffer_params_t){
+          .type = IREE_HAL_MEMORY_TYPE_HOST_LOCAL,
+          .usage =
+              IREE_HAL_BUFFER_USAGE_TRANSFER | IREE_HAL_BUFFER_USAGE_MAPPING,
+      },
+      iree_const_byte_span_empty(), dst);
+}
+
+// Allocates device-local |dst| to have the same shape as |src|.
+// Implicitly zero-filled.
+static iree_status_t allocate_device_buffer_view_like(
+    iree_hal_allocator_t* hal_allocator, iree_hal_buffer_view_t* src,
+    iree_hal_buffer_view_t** dst) {
+  return iree_hal_buffer_view_allocate_buffer(
+      hal_allocator, iree_hal_buffer_view_shape_dims(src),
+      iree_hal_buffer_view_shape_rank(src),
+      iree_hal_buffer_view_element_type(src),
+      iree_hal_buffer_view_encoding_type(src),
+      (iree_hal_buffer_params_t){
+          .type = IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL,
+          .usage =
+              IREE_HAL_BUFFER_USAGE_TRANSFER | IREE_HAL_BUFFER_USAGE_DISPATCH,
+      },
+      iree_const_byte_span_empty(), dst);
+}
+
+// Performs a deep copy of device-local |src| into host-local |dst|.
+// Allocates |dst|.
+static iree_status_t copy_device_buffer_view_to_host(
+    iree_hal_device_t* device, iree_hal_allocator_t* hal_allocator,
+    iree_hal_buffer_view_t* src, iree_hal_buffer_view_t** dst) {
+  IREE_RETURN_IF_ERROR(
+      validate_memory_type(src, IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL));
+  IREE_RETURN_IF_ERROR(allocate_host_buffer_view_like(hal_allocator, src, dst));
+  void* dst_data = NULL;
+  iree_status_t status = map_host_local_row_major_data(
+      *dst, IREE_HAL_MEMORY_ACCESS_WRITE, &dst_data);
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_device_transfer_d2h(
+        device, iree_hal_buffer_view_buffer(src), 0, dst_data,
+        iree_hal_buffer_view_byte_length(src),
+        IREE_HAL_TRANSFER_BUFFER_FLAG_DEFAULT, iree_infinite_timeout());
+  }
+  return status;
+}
+
+// Performs a deep copy of host-local |src| into a device-local |dst|.
+// Allocates |dst|.
+static iree_status_t copy_host_buffer_view_to_device(
+    iree_hal_device_t* device, iree_hal_allocator_t* hal_allocator,
+    iree_hal_buffer_view_t* src, iree_hal_buffer_view_t** dst) {
+  IREE_RETURN_IF_ERROR(
+      validate_memory_type(src, IREE_HAL_MEMORY_TYPE_HOST_LOCAL));
+  IREE_RETURN_IF_ERROR(
```

Comment:
Instead of alloc + transfer you can pass the initial buffer data directly to the allocation routine - this avoids the extra copy when possible. Instead of alloc + map + transfer you'd pass in the host pointer as `initial_data` on `iree_hal_buffer_view_allocate_buffer`:
```c
void* src_data = NULL;
map_host_local_row_major_data(..., &src_data);
iree_hal_buffer_view_allocate_buffer(..., iree_make_const_byte_span(src_data, iree_hal_buffer_view_byte_length(src), &dst);
```
(or, make your map helper return the byte span for brevity)

Reply:
(or as below, if you generate the data during allocation instead of having your own alloc + copy + etc helpers you don't need this function at all)

Reply:
Done.

Actually this made me realize that `map_host_local_row_major_data` 's signature was inherently flawed: by returning only the `void*` data pointer, it didn't give the caller a chance to later unmap the mapping. So now it returns the mapping, and we unmap those. And like suggested, the mapping byte span is passed as `initial_data` when allocating a buffer in that `copy` function, removing the need for the `h2d` transfer.